tests/test_bgt.py

```diff
@@ -13,96 +14,64 @@
-        games_raw = load(games_put_json)
-        # validate/cast game data
-    games = []
-    for game in games_raw:
-        games.append(Game(**game))
-    return games
-
-
-@fixture
-def outside_game_data():
-    with open(Path(__file__).parent/"outside_game.json", "r") as outside_game_json:
-        return Game(**load(outside_game_json))
-
-
-@fixture
-def setup_db(games_data, aws_credentials):
-    with mock_dynamodb2():
-        # load table schema
-        with open(Path(__file__).parent.parent/"games_table_schema.json", "r") as schema_json:
-            schema = load(schema_json)
-
-        # create dynamodb interface (mocked by moto decorator)
-        dynamodb = boto3.resource('dynamodb', 'us-east-1')
-
-        # create table with schema
-        table = dynamodb.create_table(
-            **schema
-        )
-        table.wait_until_exists()
-
-        # put games in table
-        with table.batch_writer() as batch:
-            for game in games_data:
-                batch.put_item(game.dict())
-        from app._main import app
-        client = TestClient(app)
-        yield client
-
-
 def test_version():
-    assert __version__ == '0.1.0'
+    with open(Path(__file__).parent.parent / "pyproject.toml") as f:
+        pyproject_toml = toml.load(f)
+    assert __version__ == pyproject_toml["tool"]["poetry"]["version"]
 
 
-def test_get_nogame(outside_game_data, setup_db):
+def test_get_nogame(outside_game_data, client):
     "Makes sure we're testing against the mocked DB and not the real one"
-    client = setup_db
     test_game = outside_game_data
     assert hasattr(test_game, "id")
     assert outside_game_data.id == 59946
     response = client.get(f"/games/{test_game.id}")
     assert response.status_code == 404
 
 
-def test_root(setup_db):
-    "The root should return the current version"
-    client = setup_db
+def test_root(client):
+    "Should return the current version"
     response = client.get("/")
-    print(response)
     assert response.status_code == 200
-    assert 0
 
 
-def test_get_game(games_data, setup_db):
-    client = setup_db
+def test_get_game(games_data, client):
+    "Should successfully get game by id"
     test_game = games_data[randint(0, 24)]
     response = client.get(f"/games/{test_game.id}")
     assert hasattr(test_game, "id")
     assert response.status_code == 200
 
 
-def test_get_games(setup_db):
-    client = setup_db
+def test_get_games(client):
+    "Should successfully get all games"
     response = client.get(f"/games/")
     result = loads(response.content)
     assert response.status_code == 200
     assert len(result) == 25
+
+def test_get_question(games_data, client):
+    "GET questions takes 1 query param, asked --"
+    "asked is a comma separated list of question ids that have been asked"
+    "memo is passed back and forth between server and client to ensure "
+    "questions are not repeated"
+    ids = set(g.id for g in games_data)

```

Comment:
This is the point at which I would consider adding a layer of abstraction around Question selection. I think I would expect this test to show, via mock instances, that when you call the questions endpoint, the QuestionSelector class is invoked. Then you would have a suite of smaller tests that you run against your QuestionSelector class with more specific assertions.

Reply:
I think I understand. Let me make sure I am planning the right changes 

1. I simplify this test to simply check that the end point
2. That I define a class in `app` that groups together code related to question selection
3. I test the rules I'm ascribing to question selection by importing that class and running tests on it directly, eliminating the TestClient layer for those tests

Reply:
That sounds right to me. You’ll have to figure out how to supply your endpoint with a mock instance during tests and a real instance in production. I use a package called mox3 to create generic mock objects that allow you to record a series of expectations and then ensure that those expectations are replayed exactly as they were recorded when you execute the function under test.

Reply:
That sounds right to me. You’ll have to figure out how to supply your endpoint with a mock instance during tests and a real instance in production. I use a package called mox3 to create generic mock objects that allow you to record a series of expectations and then ensure that those expectations are replayed exactly as they were recorded when you execute the function under test.

Reply:
took a swing here, let me know what you think